server/src/main/java/io/druid/server/coordinator/cost/SegmentCostCache.java

```diff
@@ -0,0 +1,379 @@
+        --leftIndex;
+      }
+      // add left-non-overlapping segments
+      if (leftIndex >= 0) {
+        cost += leftSum[leftIndex] * (FastMath.exp(-t1) - FastMath.exp(-t0));
+      }
+      // add all right-overlapping segments
+      int rightIndex = index;
+      while (rightIndex < sortedSegments.size() && sortedSegments.get(rightIndex)
+                                                                 .getInterval()
+                                                                 .overlaps(dataSegment.getInterval())) {
+        double start = convertStart(sortedSegments.get(rightIndex), interval);
+        double end = convertEnd(sortedSegments.get(rightIndex), interval);
+        cost += CostBalancerStrategy.intervalCost(t1 - t0, start - t0, end - t0);
+        ++rightIndex;
+      }
+      // add right-non-overlapping segments
+      if (rightIndex < sortedSegments.size()) {
+        cost += rightSum[rightIndex] * (FastMath.exp(t0) - FastMath.exp(t1));
+      }
+      return cost;
+    }
+
+    private static double convertStart(DataSegment dataSegment, Interval interval)
+    {
+      return toLocalInterval(dataSegment.getInterval().getStartMillis(), interval);
+    }
+
+    private static double convertEnd(DataSegment dataSegment, Interval interval)
+    {
+      return toLocalInterval(dataSegment.getInterval().getEndMillis(), interval);
+    }
+
+    private static double toLocalInterval(long millis, Interval interval)
+    {
+      return (millis - interval.getStartMillis()) / MILLIS_FACTOR;
+    }
+
+    public static Builder builder(Interval interval)
+    {
+      return new Builder(interval);
+    }
+
+    static class Builder
+    {
+      private final Interval interval;
+      private final LinkedList<DataSegment> dataSegments = new LinkedList<>();
+      private final SumList leftSum = new SumList(false);
+      private final SumList rightSum = new SumList(true);
+
+      public Builder(Interval interval)
+      {
+        this.interval = interval;
+      }
+
+      public void addSegment(DataSegment dataSegment)
+      {
+        if (!interval.contains(dataSegment.getInterval().getStartMillis())) {
+          throw new ISE("Failed to add segment to bucket: interval is not covered by this bucket");
+        }
+        int index = Collections.binarySearch(dataSegments, dataSegment, SEGMENT_INTERVAL_COMPARATOR);
+        index = (index >= 0) ? index : -index - 1;
+
+        dataSegments.add(index, dataSegment);
+
+        double t0 = convertStart(dataSegment, interval);
+        double t1 = convertEnd(dataSegment, interval);
+
+        double leftValue = FastMath.exp(t0) - FastMath.exp(t1);
+        double rightValue = FastMath.exp(-t1) - FastMath.exp(-t0);
+
+        leftSum.add(index, leftValue);
+        rightSum.add(index, rightValue);
+      }
+
+      public void removeSegment(DataSegment dataSegment)
+      {
+        int index = -1;
+        ListIterator<DataSegment> it = dataSegments.listIterator();
```

Comment:
Please define `it` in the loop scope. Also it's type don't need to be `ListIterator`, could be just iterator.